[GeoMechanicsApplication] Avoid division by zero when calculating Lode angle

avdg81 opened the pull request:
**📝 Description**
When calculating Lode angle, avoid division by zero. When the Von Mises stress equals (nearly) zero, Lode angle should be zero, too.


**🆕 Changelog**
Other changes include:
- No special handling is needed when the `numerator` equals (nearly) zero.
- The calculation of Lode angle first checks for the special case where the Von Mises stress equals (nearly) zero. Only calculate the principal stresses and eigenvectors when they are needed, rather than upfront.
- Avoid a domain error raised by `std::asin` by clamping the input value to [-1, +1].
- Reuse a vector instance in the unit test for the calculation of Lode angle.
- Added some comments to the unit test to make clear that all relevant cases (i.e. triaxial extension [TXE], triaxial compression [TXC], and shear [SHR]) have been covered.

avdg81 committed:
Refactored the calculation of Lode angle

Also avoid division by zero, by checking the Von Mises stress.

avdg81 committed:
Removed an unneeded test variable

Also commented on the performed validations.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -64,14 +64,16 @@ double StressStrainUtilities::CalculateLodeAngle(const Vector& rStressVector)
-    const double numerator = (sigma_princi(0, 0) - p) * (sigma_princi(1, 1) - p) * (sigma_princi(2, 2) - p);
-    if (std::abs(numerator) < 1.0E-12) return 0.;
+    const auto q = CalculateVonMisesStress(rStressVector);
+    if (constexpr auto tolerance = 1.0E-12; q < tolerance) return 0.0;
```
Shouldn't we use something like Defaults::absolute_tolerance?
The 1.0e-16 that I managed to put on line 72 is equally bad.

avdg81 replied:
The defaults that you are referring to are defined in `applications\GeoMechanicsApplication\tests\cpp_tests\test_utilities.h`. Since those are default tolerance values for tests, I'd rather not mix them with default tolerance values for production code. We could create a new file where we define application-wide tolerance values, and use that here. Let me know whether that is what you've intended to say. Thanks.

WPK4FEM replied:
I do not want to cause scope creep of the PR. Leave as is.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_utilities/test_stress_strain_utitlities.cpp:
```diff
@@ -58,15 +58,17 @@ KRATOS_TEST_CASE_IN_SUITE(CheckCalculateVonMisesStressPureShear, KratosGeoMechan
-    Vector stress_vector2(4);
-    stress_vector2 <<= -1.0, 0.0, 1.0, 0.0;
+    // Validate Shear (SHR)
+    stress_vector <<= -1.0, 0.0, 1.0, 0.0;
```
No clue why I introduced an extra variable before. This is clearer.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/stress_strain_utilities.cpp:
```diff
@@ -64,14 +64,16 @@ double StressStrainUtilities::CalculateLodeAngle(const Vector& rStressVector)
+    const auto p = CalculateMeanStress(rStressVector);
+    const auto numerator = (sigma_princi(0, 0) - p) * (sigma_princi(1, 1) - p) * (sigma_princi(2, 2) - p);
+
     return std::asin((-27. / 2.) * numerator / (q * q * q)) / 3.0;
```
As a first fix we also used to test the argument of the asin (`(-27. / 2.) * numerator / (q * q * q)`) is not just over 1/ under -1 (within some tolerance). Maybe we can reproduce it if we slightly perturb the vectors in the UT to trigger it (e.g. get the argument to be 1.0 + 1e-12)?

avdg81 replied:
Thank you very much for your suggestion. I have played a bit with finding the right perturbation, and when I found it I've added two regression tests for it. The value of the number passed to `std::asin` is now `std::clamp`ed to the domain [-1, +1].

rfaasse replied:
Nice find! Thanks for processing this!

rfaasse reviewed (commented):
No blocking comment, just a suggestion for reproducing one of the earlier things we encountered

avdg81 committed:
Avoid domain errors when calling `std::asin`

Also added two regression tests that show clamping the argument of `std::asin` to [-1.0, 1.0] works.

rfaasse reviewed (approved):
Thanks for processing the review comments, in my opinion this is ready to go!
